### PR TITLE
FINERACT-38

### PR DESCRIPTION
**Pull request by maektwain**
Test Integrated

Improved Tests

Adding PreviewIncomeAndExpense

## Description
Describe the changes made and why they were made. Ignore if these details are present on the associated Jira ticket

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Commit message starts with the issue number from https://issues.apache.org/jira/projects/FINERACT/. Ex: FINERACT-646 Pockets API.

- [x] Coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions have been followed.

- [x] API documentation at https://github.com/apache/fineract/blob/develop/api-docs/apiLive.htm has been updated with details of any API changes.

- [x] Integration tests have been created/updated for verifying the changes made.

- [x] All Integrations tests are passing with the new commits.

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the list.)

Our guidelines for code reviews is at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide

### CODE REVIEW

**Comment by maektwain** [comment]
@vorburger Let's see, if this fails, at my end it passed.

**Comment by maektwain** [comment]
@xurror could you see and get this reviewed ?

File: fineract-provider/src/main/java/org/apache/fineract/organisation/office/service/OfficeReadPlatformServiceImpl.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/organisation/office/service/OfficeReadPlatformServiceImpl.java]: safer to return a `Collections.emptyList()` instead of `null` here

File: fineract-provider/src/main/java/org/apache/fineract/accounting/closure/api/GLClosuresApiResource.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/closure/api/GLClosuresApiResource.java]: safer to use a `Collections.emptySet()` instead of `new HashSet<String>()` here

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java]: This confused me while I was doing a technical code review of this PR, so I had a closer look at the existing code when I realized that this was just copy/pasted... it's already wrong in that existing code, I'm proposing to improve it in https://github.com/apache/fineract/pull/799.. would you be willing to rebase and change this accordingly, once that is merged?

File: fineract-provider/src/main/java/org/apache/fineract/accounting/closure/service/GLClosureWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/closure/service/GLClosureWritePlatformServiceJpaRepositoryImpl.java]: This `return null`here looks very curious... won't this hide problems, are we sure that the (commented out!) throwing of an exception wouldn't be better?

**Review by vorburger — CHANGES_REQUESTED** [review]
I have done a quick technical review, and have some minor feedback.

Other than that, someone else should probably do a more functional review? 

We could also just merge after e.g. 2 weeks "grace period" if nobody can a functional review within 2 weeks after technical feedback is taken into account (we have done this before for other functional enhancements).

**Comment by maektwain** [comment]
> I have done a quick technical review, and have some minor feedback.
> 
> Other than that, someone else should probably do a more functional review?

Should I qualify as "other ", since I have done end to end functional review and testing. Though we need some "meta-other" to kind of validate "other" (Waiting)
> 
> We could also just merge after e.g. 2 weeks "grace period" if nobody can a functional review within 2 weeks after technical feedback is taken into account (we have done this before for other functional enhancements).

I think we should write some functional tests also (We could cover such scenario, then we won't be much dependent on others to validate the functional test) then in the end only manual testing is required?

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java
Comment (maektwain) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java]: Rebase here ack.

File: fineract-provider/src/main/java/org/apache/fineract/accounting/closure/service/GLClosureWritePlatformServiceJpaRepositoryImpl.java
Comment (maektwain) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/closure/service/GLClosureWritePlatformServiceJpaRepositoryImpl.java]: I did change it to throw.

File: fineract-provider/src/main/java/org/apache/fineract/accounting/closure/api/GLClosuresApiResource.java
Comment (maektwain) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/closure/api/GLClosuresApiResource.java]: Ack.

**Comment by maektwain** [comment]
@vorburger This test failed 

```
org.apache.fineract.integrationtests.SchedulerJobsTestResults > testApplyHolidaysToLoansJobOutcome FAILED
8245    java.lang.AssertionError at SchedulerJobsTestResults.java:326
```

**Comment by vorburger** [comment]
@maektwain see https://github.com/apache/fineract#pull-requests Basically, do not wait for an active maintainer to come and help you re. build failures - it's your responsibility to get a green build. Have you done what the README suggests to do in such cases? 😺

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpenseBookingIntegrationTest.java
Comment (vorburger) [inline fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpenseBookingIntegrationTest.java]: Isn't this actually a Unit and not an Integration Test? It doesn't seem to make any REST calls (as far as I can tell). If that is correct, then how about moving this to `src/test` instead of `src/integregationTest`?

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpenseBookingIntegrationTest.java
Comment (vorburger) [inline fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpenseBookingIntegrationTest.java]: This (`@Test(expected = NoClassDefFoundError.class)`) seems really weird to me, at first review. Not the expecting an exception, that is fine, as above in `testBookOffIncomeAndExpenseCheckException` for `RunningBalanceNotCalculatedException`, but why would a "functional" test ever hit a `NoClassDefFoundError`, which is a (very) low-level technical exception that (normally) "should never happen"?

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpenseBookingIntegrationTest.java
Comment (vorburger) [inline fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpenseBookingIntegrationTest.java]: as above

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpenseBookingIntegrationTest.java
Comment (vorburger) [inline fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpenseBookingIntegrationTest.java]: again

File: fineract-provider/src/main/java/org/apache/fineract/accounting/closure/domain/IncomeAndExpenseBooking.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/closure/domain/IncomeAndExpenseBooking.java:48]: Minor formatting nit pick: How about a blank line space between each annotated JPA field?

File: fineract-provider/src/main/java/org/apache/fineract/accounting/closure/service/CalculateIncomeAndExpenseBookingImpl.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/closure/service/CalculateIncomeAndExpenseBookingImpl.java]: ?

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java]: but it's still done the old one, didn't you mean to change this, after the rebase?

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java]:
```suggestion
            throw handleJournalEntryDataIntegrityIssues(dve);
```

**Review by vorburger — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java
Comment (maektwain) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java]: I rebased it , it did not push here, 🥇

File: fineract-provider/src/main/java/org/apache/fineract/accounting/closure/service/CalculateIncomeAndExpenseBookingImpl.java
Comment (maektwain) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/closure/service/CalculateIncomeAndExpenseBookingImpl.java]: https://github.com/muellners/fineract/blob/38e19f73fcea5c9165d34d9c83a7be8bf6cec884/fineract-provider/src/main/java/org/apache/fineract/accounting/closure/service/GLClosureWritePlatformServiceJpaRepositoryImpl.java#L353

Kinda of strange , in my repo I updated this, but does it not reflect here,

Comment (maektwain) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/closure/service/CalculateIncomeAndExpenseBookingImpl.java]: You are looking at the old revision @vorburger

**Comment by maektwain** [comment]
Hey @vorburger could you make this as draft, since I tested the test written they don't comply and need to re-write them again. (real integration test)

**Comment by vorburger** [comment]
@maektwain friendly ping / reminder - are you still planning to work on finishing this one?

**Comment by maektwain** [comment]
Yeah , I will put some time on the pending integration test. Give me couple of days,

File: fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/accounting/journalentry/service/JournalEntryWritePlatformServiceJpaRepositoryImpl.java]: Resolving because this was actually done meanwhile (@maektwain in the future please comment to let reviewer know).

**Comment by vorburger** [comment]
@maektwain friendly ping / reminder - are you still planning to work on finishing this one? Rebase it.

**Comment by maektwain** [comment]
Yeah , @vorburger I am planning to finish this, I need to write tests which are really crazy will be able  to finish this tomorrow

**Pull request closed without merging**
